# Explicit ABBA guiding-center symplecticity study

This experiment evaluates the physical guiding-center map produced by `ExplicitABBA`. At the beginning of every step, the physical state is duplicated, $E y_n=(y_n,y_n)$. For $s=h/2$, the copies follow $A_{s,t_n}$, $B_{s,t_n}$, $B_{s,t_n+h}$, and $A_{s,t_n+h}$, and the final physical state is their arithmetic mean:

$$y_{n+1}=P\Phi^{\mathrm{ABBA}}_{h,t_n}E y_n,\qquad P(u,v)=\frac{u+v}{2}. $$

The method is explicit and second order, but $P$ is a Euclidean projection and does not guarantee symplecticity. For every complete step, the experiment differentiates the emitted physical map by centered finite differences. If $J_n$ is the local step Jacobian and $DG_n$ is the accumulated discrete-flow Jacobian, it records

$$\varepsilon_{\mathrm{local},n}=\frac{\|J_n^T\Omega J_n-\Omega\|_F}{\|\Omega\|_F},\qquad \varepsilon_{\mathrm{flow},n}=\frac{\|DG_n^T\Omega DG_n-\Omega\|_F}{\|\Omega\|_F}. $$

Transported polygon area, $|\det(DG_n)-1|$, and the pre-projection separation $\|u^{(2)}-v^{(2)}\|_\infty$ provide complementary geometric diagnostics.

In [1]:
import numpy as np

from studies import (
    ExplicitABBASymplecticityConfig,
    RandomPotentialConfig,
    centered_circle,
    pi_area_steps,
    run_explicit_abba_symplecticity_study,
)

## Reproducible configuration

The potential, initial boundary, physical parameters, time interval, ABBA steps, observation grid, and finite-difference scale are explicit below. The random field, radius, and $\rho$ match the RK4 and symmetric-projected ABBA experiments in this directory. Eight boundary points and the interval $[0,\pi]$ keep the full step-by-step Jacobian propagation practical while retaining three temporal refinements.

In [2]:
potential_config = RandomPotentialConfig(
    amplitude=0.7,
    max_wave_number=25,
    nx=64,
    ny=64,
    seed=27,
    interpolation_order=5,
)
potential = potential_config.build()

circle_radius = 0.5
circle_points = 8
rho = 0.3
circle = centered_circle(
    potential,
    radius=circle_radius,
    points=circle_points,
    rho=rho,
)

finite_difference_relative_step = float(np.cbrt(np.finfo(float).eps))
study_config = ExplicitABBASymplecticityConfig(
    steps=pi_area_steps(20, 40, 80),
    t_span=(0.0, np.pi),
    save_interval=np.pi / 10,
    chunk_size=16,
    progress=True,
    block_prefix="explicit_abba_symplecticity",
    finite_difference_relative_step=finite_difference_relative_step,
)

print(potential_config)
print(study_config)
print(
    f"Circle: {circle_points} points; t={study_config.t_span}; "
    f"{study_config.output_sample_count} saved states; "
    f"finite-difference relative step={finite_difference_relative_step:.8e}"
)

RandomPotentialConfig(amplitude=0.7, max_wave_number=25, nx=64, ny=64, seed=27, interpolation_order=5)
ExplicitABBASymplecticityConfig(steps=(AreaStep(label='$\\Delta t=\\pi/20$', value=0.15707963267948966), AreaStep(label='$\\Delta t=\\pi/40$', value=0.07853981633974483), AreaStep(label='$\\Delta t=\\pi/80$', value=0.039269908169872414)), t_span=(0.0, 3.141592653589793), save_interval=0.3141592653589793, rho=None, chunk_size=16, progress=True, block_prefix='explicit_abba_symplecticity', finite_difference_relative_step=6.0554544523933395e-06)
Circle: 8 points; t=(0.0, 3.141592653589793); 11 saved states; finite-difference relative step=6.05545445e-06


In [3]:
finite_difference_relative_step

6.0554544523933395e-06

## Explicit ABBA integrations and persisted Jacobians

Every main-grid ABBA step contributes to the accumulated finite-difference Jacobian. Scalar diagnostics and full local and accumulated Jacobians are persisted at the common observation times. Shadow advances used only for output interpolation do not affect the trajectory or the diagnostic history.

In [4]:
result = run_explicit_abba_symplecticity_study(
    potential,
    circle,
    notebook_path=(
        "notebooks/experiments/symplecticity/"
        "gc_area_and_explicit_abba_symplecticity.ipynb"
    ),
    config=study_config,
    metadata={
        **potential_config.metadata(),
        "circle_radius": circle_radius,
        "study_kind": "versioned_symplecticity_experiment",
    },
)
result.print_summary()

ExplicitABBA [=>                            ]   5.0% (1/20, t=0.15708)

ExplicitABBA [===>                          ]  10.0% (2/20, t=0.314159)

ExplicitABBA [====>                         ]  15.0% (3/20, t=0.471239)

ExplicitABBA [======>                       ]  20.0% (4/20, t=0.628319)

ExplicitABBA [=======>                      ]  25.0% (5/20, t=0.785398)

ExplicitABBA [=========>                    ]  30.0% (6/20, t=0.942478)

ExplicitABBA [==========>                   ]  35.0% (7/20, t=1.09956)

ExplicitABBA [============>                 ]  40.0% (8/20, t=1.25664)

ExplicitABBA [=============>                ]  45.0% (9/20, t=1.41372)

ExplicitABBA [===============>              ]  50.0% (10/20, t=1.5708)

ExplicitABBA [================>             ]  55.0% (11/20, t=1.72788)

ExplicitABBA [==================>           ]  60.0% (12/20, t=1.88496)

ExplicitABBA [===================>          ]  65.0% (13/20, t=2.04204)

ExplicitABBA [=====================>        ]  70.0% (14/20, t=2.19911)

ExplicitABBA [======================>       ]  75.0% (15/20, t=2.35619)

ExplicitABBA [========================>     ]  80.0% (16/20, t=2.51327)

ExplicitABBA [=========================>    ]  85.0% (17/20, t=2.67035)

ExplicitABBA [===========================>  ]  90.0% (18/20, t=2.82743)

ExplicitABBA [============================> ]  95.0% (19/20, t=2.98451)

ExplicitABBA [==============================] 100.0% (20/20, t=3.14159)

ExplicitABBA [>                             ]   2.5% (1/40, t=0.0785398)

ExplicitABBA [=>                            ]   5.0% (2/40, t=0.15708)

ExplicitABBA [==>                           ]   7.5% (3/40, t=0.235619)

ExplicitABBA [===>                          ]  10.0% (4/40, t=0.314159)

ExplicitABBA [===>                          ]  12.5% (5/40, t=0.392699)

ExplicitABBA [====>                         ]  15.0% (6/40, t=0.471239)

ExplicitABBA [=====>                        ]  17.5% (7/40, t=0.549779)

ExplicitABBA [======>                       ]  20.0% (8/40, t=0.628319)

ExplicitABBA [======>                       ]  22.5% (9/40, t=0.706858)

ExplicitABBA [=======>                      ]  25.0% (10/40, t=0.785398)

ExplicitABBA [========>                     ]  27.5% (11/40, t=0.863938)

ExplicitABBA [=========>                    ]  30.0% (12/40, t=0.942478)

ExplicitABBA [=========>                    ]  32.5% (13/40, t=1.02102)

ExplicitABBA [==========>                   ]  35.0% (14/40, t=1.09956)

ExplicitABBA [===========>                  ]  37.5% (15/40, t=1.1781)

ExplicitABBA [============>                 ]  40.0% (16/40, t=1.25664)

ExplicitABBA [============>                 ]  42.5% (17/40, t=1.33518)

ExplicitABBA [=============>                ]  45.0% (18/40, t=1.41372)

ExplicitABBA [==============>               ]  47.5% (19/40, t=1.49226)

ExplicitABBA [===============>              ]  50.0% (20/40, t=1.5708)

ExplicitABBA [===============>              ]  52.5% (21/40, t=1.64934)

ExplicitABBA [================>             ]  55.0% (22/40, t=1.72788)

ExplicitABBA [=================>            ]  57.5% (23/40, t=1.80642)

ExplicitABBA [==================>           ]  60.0% (24/40, t=1.88496)

ExplicitABBA [==================>           ]  62.5% (25/40, t=1.9635)

ExplicitABBA [===================>          ]  65.0% (26/40, t=2.04204)

ExplicitABBA [====================>         ]  67.5% (27/40, t=2.12058)

ExplicitABBA [=====================>        ]  70.0% (28/40, t=2.19911)

ExplicitABBA [=====================>        ]  72.5% (29/40, t=2.27765)

ExplicitABBA [======================>       ]  75.0% (30/40, t=2.35619)

ExplicitABBA [=======================>      ]  77.5% (31/40, t=2.43473)

ExplicitABBA [========================>     ]  80.0% (32/40, t=2.51327)

ExplicitABBA [========================>     ]  82.5% (33/40, t=2.59181)

ExplicitABBA [=========================>    ]  85.0% (34/40, t=2.67035)

ExplicitABBA [==========================>   ]  87.5% (35/40, t=2.74889)

ExplicitABBA [===========================>  ]  90.0% (36/40, t=2.82743)

ExplicitABBA [===========================>  ]  92.5% (37/40, t=2.90597)

ExplicitABBA [============================> ]  95.0% (38/40, t=2.98451)

ExplicitABBA [=============================>]  97.5% (39/40, t=3.06305)

ExplicitABBA [==============================] 100.0% (40/40, t=3.14159)

ExplicitABBA [>                             ]   1.2% (1/80, t=0.0392699)

ExplicitABBA [>                             ]   2.5% (2/80, t=0.0785398)

ExplicitABBA [=>                            ]   3.8% (3/80, t=0.11781)

ExplicitABBA [=>                            ]   5.0% (4/80, t=0.15708)

ExplicitABBA [=>                            ]   6.2% (5/80, t=0.19635)

ExplicitABBA [==>                           ]   7.5% (6/80, t=0.235619)

ExplicitABBA [==>                           ]   8.8% (7/80, t=0.274889)

ExplicitABBA [===>                          ]  10.0% (8/80, t=0.314159)

ExplicitABBA [===>                          ]  11.2% (9/80, t=0.353429)

ExplicitABBA [===>                          ]  12.5% (10/80, t=0.392699)

ExplicitABBA [====>                         ]  13.8% (11/80, t=0.431969)

ExplicitABBA [====>                         ]  15.0% (12/80, t=0.471239)

ExplicitABBA [====>                         ]  16.2% (13/80, t=0.510509)

ExplicitABBA [=====>                        ]  17.5% (14/80, t=0.549779)

ExplicitABBA [=====>                        ]  18.8% (15/80, t=0.589049)

ExplicitABBA [======>                       ]  20.0% (16/80, t=0.628319)

ExplicitABBA [======>                       ]  21.2% (17/80, t=0.667588)

ExplicitABBA [======>                       ]  22.5% (18/80, t=0.706858)

ExplicitABBA [=======>                      ]  23.8% (19/80, t=0.746128)

ExplicitABBA [=======>                      ]  25.0% (20/80, t=0.785398)

ExplicitABBA [=======>                      ]  26.2% (21/80, t=0.824668)

ExplicitABBA [========>                     ]  27.5% (22/80, t=0.863938)

ExplicitABBA [========>                     ]  28.7% (23/80, t=0.903208)

ExplicitABBA [=========>                    ]  30.0% (24/80, t=0.942478)

ExplicitABBA [=========>                    ]  31.2% (25/80, t=0.981748)

ExplicitABBA [=========>                    ]  32.5% (26/80, t=1.02102)

ExplicitABBA [==========>                   ]  33.8% (27/80, t=1.06029)

ExplicitABBA [==========>                   ]  35.0% (28/80, t=1.09956)

ExplicitABBA [==========>                   ]  36.2% (29/80, t=1.13883)

ExplicitABBA [===========>                  ]  37.5% (30/80, t=1.1781)

ExplicitABBA [===========>                  ]  38.8% (31/80, t=1.21737)

ExplicitABBA [============>                 ]  40.0% (32/80, t=1.25664)

ExplicitABBA [============>                 ]  41.2% (33/80, t=1.29591)

ExplicitABBA [============>                 ]  42.5% (34/80, t=1.33518)

ExplicitABBA [=============>                ]  43.8% (35/80, t=1.37445)

ExplicitABBA [=============>                ]  45.0% (36/80, t=1.41372)

ExplicitABBA [=============>                ]  46.2% (37/80, t=1.45299)

ExplicitABBA [==============>               ]  47.5% (38/80, t=1.49226)

ExplicitABBA [==============>               ]  48.8% (39/80, t=1.53153)

ExplicitABBA [===============>              ]  50.0% (40/80, t=1.5708)

ExplicitABBA [===============>              ]  51.2% (41/80, t=1.61007)

ExplicitABBA [===============>              ]  52.5% (42/80, t=1.64934)

ExplicitABBA [================>             ]  53.8% (43/80, t=1.68861)

ExplicitABBA [================>             ]  55.0% (44/80, t=1.72788)

ExplicitABBA [================>             ]  56.2% (45/80, t=1.76715)

ExplicitABBA [=================>            ]  57.5% (46/80, t=1.80642)

ExplicitABBA [=================>            ]  58.8% (47/80, t=1.84569)

ExplicitABBA [==================>           ]  60.0% (48/80, t=1.88496)

ExplicitABBA [==================>           ]  61.3% (49/80, t=1.92423)

ExplicitABBA [==================>           ]  62.5% (50/80, t=1.9635)

ExplicitABBA [===================>          ]  63.7% (51/80, t=2.00277)

ExplicitABBA [===================>          ]  65.0% (52/80, t=2.04204)

ExplicitABBA [===================>          ]  66.2% (53/80, t=2.08131)

ExplicitABBA [====================>         ]  67.5% (54/80, t=2.12058)

ExplicitABBA [====================>         ]  68.8% (55/80, t=2.15984)

ExplicitABBA [=====================>        ]  70.0% (56/80, t=2.19911)

ExplicitABBA [=====================>        ]  71.2% (57/80, t=2.23838)

ExplicitABBA [=====================>        ]  72.5% (58/80, t=2.27765)

ExplicitABBA [======================>       ]  73.8% (59/80, t=2.31692)

ExplicitABBA [======================>       ]  75.0% (60/80, t=2.35619)

ExplicitABBA [======================>       ]  76.2% (61/80, t=2.39546)

ExplicitABBA [=======================>      ]  77.5% (62/80, t=2.43473)

ExplicitABBA [=======================>      ]  78.8% (63/80, t=2.474)

ExplicitABBA [========================>     ]  80.0% (64/80, t=2.51327)

ExplicitABBA [========================>     ]  81.2% (65/80, t=2.55254)

ExplicitABBA [========================>     ]  82.5% (66/80, t=2.59181)

ExplicitABBA [=========================>    ]  83.8% (67/80, t=2.63108)

ExplicitABBA [=========================>    ]  85.0% (68/80, t=2.67035)

ExplicitABBA [=========================>    ]  86.2% (69/80, t=2.70962)

ExplicitABBA [==========================>   ]  87.5% (70/80, t=2.74889)

ExplicitABBA [==========================>   ]  88.8% (71/80, t=2.78816)

ExplicitABBA [===========================>  ]  90.0% (72/80, t=2.82743)

ExplicitABBA [===========================>  ]  91.2% (73/80, t=2.8667)

ExplicitABBA [===========================>  ]  92.5% (74/80, t=2.90597)

ExplicitABBA [============================> ]  93.8% (75/80, t=2.94524)

ExplicitABBA [============================> ]  95.0% (76/80, t=2.98451)

ExplicitABBA [============================> ]  96.2% (77/80, t=3.02378)

ExplicitABBA [=============================>]  97.5% (78/80, t=3.06305)

ExplicitABBA [=============================>]  98.8% (79/80, t=3.10232)

ExplicitABBA [==============================] 100.0% (80/80, t=3.14159)

                  step           ExplicitABBA steps     max |area error|     max local defect      max flow defect      max |det-1|
     $\Delta t=\pi/20$                           20       1.36678868e-02       1.06369674e-06       2.38900107e-06   4.64027783e-06
     $\Delta t=\pi/40$                           40       1.37232101e-02       1.91463342e-08       8.88536101e-08   2.20605237e-07
     $\Delta t=\pi/80$                           80       1.37369504e-02       2.53123939e-10       2.69723344e-09   1.03814780e-08

Maximum off-diagonal copy separation before averaging:
  $\Delta t=\pi/20$: 4.83370287e-04
  $\Delta t=\pi/40$: 9.38302262e-05
  $\Delta t=\pi/80$: 1.28780677e-05

Empirical orders (local defect / flow defect / copy separation):
  $\Delta t=\pi/20$ -> $\Delta t=\pi/40$: 5.795875 / 4.748833 / 2.365004
  $\Delta t=\pi/40$ -> $\Delta t=\pi/80$: 6.241080 / 5.041878 / 2.865137


## Time-dependent diagnostics

The local matrix defect tests each averaged ABBA step independently. The accumulated defect and determinant error test the complete discrete flow from the initial boundary, while the area panel measures the transported polygon geometry.

In [5]:
diagnostic_figure, diagnostic_axes = result.plot_diagnostics()

## Defect and copy-separation convergence

Unlike the symmetric-projected ABBA method, the explicit averaged map is not expected to reach only a numerical symplecticity floor. Its local and accumulated physical defects should decrease as the integration step is refined. The copy separation measures the size of the arithmetic projection applied at each step.

In [6]:
convergence_figure, convergence_axis = result.plot_convergence()
result.convergence_orders()

(ExplicitABBADefectOrder(coarse_label='$\\Delta t=\\pi/20$', fine_label='$\\Delta t=\\pi/40$', local_defect=5.795874883895077, flow_defect=4.7488332957116555, copy_separation=2.365004143504135),
 ExplicitABBADefectOrder(coarse_label='$\\Delta t=\\pi/40$', fine_label='$\\Delta t=\\pi/80$', local_defect=6.241080431346871, flow_defect=5.041878099891084, copy_separation=2.865136606423682))

## Interpretation

The duplicated ABBA composition is symplectic in its extended space, but arithmetic averaging is rank-reducing there and does not transfer that guarantee to the physical map. The quadratic test case in `docs/tex/ABBA_explicit/ABBA_explicit.tex` makes the loss exact: $\det K_h-1=h^6/64$. The random-field experiment shows the same asymptotic structure: the measured local-defect orders are $5.80$ and $6.24$, while accumulation over $O(h^{-1})$ steps lowers the flow-defect orders to $4.75$ and $5.04$. The copy-separation orders, $2.37$ and $2.87$, approach the expected cubic scaling. The maximum flow defect falls from $2.39\times10^{-6}$ to $2.70\times10^{-9}$ across the three refinements. In contrast, the polygon-area error remains near $1.37\times10^{-2}$ because the eight-vertex spatial representation dominates this temporal refinement; it should not be interpreted as failure of the observed defect convergence.